### ISSUE TITLE
Missing 1 / (Cd) normalization term?

### ISSUE THREAD

**Issue by eiphy** [Yujun-Shi/CwD#2]
Hi @Yujun-Shi,
From the paper:
![image](https://user-images.githubusercontent.com/46335499/167760433-04397604-8fa3-4904-beb3-dacb32518052.png)
In code:
/d^2
https://github.com/Yujun-Shi/CwD/blob/ac6d4eda9293fb8e0b5f39943795a6a2e264d1fd/src/approach/aux_loss.py#L49
/N
https://github.com/Yujun-Shi/CwD/blob/ac6d4eda9293fb8e0b5f39943795a6a2e264d1fd/src/approach/aux_loss.py#L57
*eta
https://github.com/Yujun-Shi/CwD/blob/ac6d4eda9293fb8e0b5f39943795a6a2e264d1fd/src/approach/lucir_cwd.py#L268

Is 1 / C missing from the code or it should be `loss = loss / uniq_l.numel()`?

**Comment by eiphy** [comment]
Besides, shouldn't:
https://github.com/Yujun-Shi/CwD/blob/ac6d4eda9293fb8e0b5f39943795a6a2e264d1fd/src/approach/aux_loss.py#L43
be `corr_mat = torch.matmul(x_label.t(), x_label) / (N - 1)`?

**Comment by Yujun-Shi** [comment]
@eiphy Hi, thanks for your interests in our work and sorry for the late response!

Essentially, Eqn. (8) in the main paper is obviously the most natural way of defining our objective and we do observe similar improvements by implementing in the exact form as Eqn. (8). However, we find that implementing exactly as in Eqn. (8) is not very numerically stable. We conjecture that it is because within one batch of data, number of samples of each class might be very different, which means directly normalize by number of instances of each class might not be the most ideal operation. Therefore, we slightly adjust the normalization here, and the results become slightly more stable. **All results in the experimental sections are produced by the version used in our code.**

However, we would like to emphasize that the spirit of our method is not in these normalization (such constant normalizations only have limited impact on performance). The determinant factor of our CwD is the decorrelation operation. As long as the representations are decorrelated, and representations don't lie in the collapsed subspace, the performance should be improved.

Additionally, the reason why we only extract the off-diagonal terms for regularization instead of penalizing the Frobenius norm of the correlation matrix is elaborated in the comment of our code.

Hope this answer can help resolve your confusion :)

**Comment by eiphy** [comment]
@Yujun-Shi 
Thank you for your explanation!
